graphql_compiler/compiler/ir_lowering_cypher/ir_lowering.py

```diff
@@ -0,0 +1,185 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+from functools import partial
+
+from ..blocks import CoerceType, Filter, Fold, MarkLocation, Recurse, Traverse
+from ..expressions import (
+    BinaryComposition, ContextField, FoldedContextField, LocalField, NullLiteral
+)
+from ..helpers import get_only_element_from_collection
+from ..ir_lowering_common.common import merge_consecutive_filter_clauses
+from ..ir_lowering_common.location_renaming import (
+    make_location_rewriter_visitor_fn, make_revisit_location_translations
+)
+from ..helpers import FoldScopeLocation
+
+
+##################################
+# Optimization / lowering passes #
+##################################
+
+def insert_explicit_type_bounds(ir_blocks, query_metadata_table, type_equivalence_hints=None):
+    """Add a CoerceType block after every Traverse/Fold/Recurse, to hint to the Cypher scheduler."""
```

Comment:
How about adding a `QueryRoot` if the root type can be inferred from an outgoing edge?

Reply:
I think that might end up being quite confusing, since that's not really what `QueryRoot` blocks are meant for.